src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -643,55 +626,166 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                // It's important to have a consistent type here, as otherwise the type of the caching field could
+                // end up changing non-deterministically based on which type for a given blob was encountered first.
+                // Also, even if we're not dealing with an enum, we still create a new array type that drops any
+                // annotations that may have initially been associated with the element type; this is similarly to
+                // ensure deterministic behavior.
+                arrayType = arrayType.WithElementType(TypeWithAnnotations.Create(elementType.EnumUnderlyingTypeOrSelf()));
+
+                var cachingField = _builder.module.GetArrayCachingFieldForData(data, _module.Translate(arrayType), wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                var arrayNotNullLabel = new object();
+
+                // T[]? array = PrivateImplementationDetails.cachingField;
+                // if (array is not null) goto arrayNotNull;
+                _builder.EmitOpCode(ILOpCode.Ldsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitBranch(ILOpCode.Brtrue, arrayNotNullLabel);
+
+                // array = new T[elementCount];
+                // RuntimeHelpers.InitializeArray(token, array);
+                // PrivateImplementationDetails.cachingField = array;
+                _builder.EmitOpCode(ILOpCode.Pop);
+                _builder.EmitIntConstant(elementCount);
+                _builder.EmitOpCode(ILOpCode.Newarr);
+                EmitSymbolToken(arrayType.ElementType, wrappedExpression.Syntax);
+                _builder.EmitArrayBlockInitializer(data, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitOpCode(ILOpCode.Stsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+
+                // arrayNotNullLabel:
+                // new ReadOnlySpan<T>(array)
+                _builder.MarkLabel(arrayNotNullLabel);
+                _builder.EmitOpCode(ILOpCode.Newobj, 0);
+                EmitSymbolToken(rosArrayCtor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                return true;
+            }
+
+            // Emit: new ReadOnlySpan<ElementType>(PrivateImplementationDetails.cachingField ??= new ElementType[] { ... constants ... })
+            bool tryEmitAsCachedArrayOfConstants(BoundArrayCreation arrayCreation, ArrayTypeSymbol arrayType, TypeSymbol elementType, NamedTypeSymbol spanType, bool used, BoundExpression? inPlaceTarget, out bool avoidInPlace)
+            {
+                avoidInPlace = false;
+
+                if (elementType.IsReferenceType && elementType.SpecialType != SpecialType.System_String)
+                {
+                    return false;
```

Comment:
>return false; [](http://example.com/codeflow?start=20&length=13)

Are we going to get here for `object`? Is this intentional?

Reply:
Yes, we get here for `object` (see `ReadOnlySpanFromArrayOfConstants_Null`) and it's intentional